### ISSUE TITLE
Prepare contract for future upgrades

### ISSUE THREAD

**Issue by LNow** [citycoins/citycoin#84]
We are doing our bests to keep our work at the highest level we might face in the future a need to upgrade contract due to bug discovery or to add or modify some of the features. Contracts needs to be prepared for such events and we need to make sure, that upgrade process won't disturb mining nor stacking process and minimize eventual the downtime.

Once we extract token code into separate contract (#83) I don't find any data other than configuration that needs to me moved to newer contract version, thus upgrading should be much easier. We do however prevent mining and stacking using old contract post upgrade and at the same time make sure that users who mined and stacked using old contract will be able to claim their rewards even many hours/days/weeks after upgrade.

To satisfy what described above we could introduce some sort of graceful shutdown or freezing process, which when initiated would set a limit (block height) that when reached would lock mining/stacking and from that point only claiming rewards would be possible.

Because stacking can be spread over a very long period of time (many cycles) and we want to let stackers stack without prolonged breaks, therefore we must also introduce a possibility to re-claim tokens they decided to stack in a future cycles that hasn't started yet and won't be active in old contract.

Whole freezing/shutdown and starting up new version of a contract could look as on the picture below:
![image](https://user-images.githubusercontent.com/829210/121260662-e7e14980-c8b1-11eb-9f02-c978fca084d6.png)

**Comment by whoabuddy** [comment]
Linking to #7 in case there is any overlap in information.

**Comment by whoabuddy** [comment]
I'm all for this - I think an upgrade path is crucial and thinking about this now rather than later is very helpful.

Would it be possible to have miners signal the upgrade through the memo field on their transactions? That way we can publish an update, but miners have to adopt it in order to activate the freeze/warmup process.

Thoughts?

**Comment by LNow** [comment]
At first I had a feeling it might be impossible to implement it using memo but I was wrong.

We can use last/first byte (or multiple bytes) for signaling miners willingness and readiness to switch to new contract.
Only when upgrade voting is active this "magic byte" is compared with byte set by contract deployer when he/she activated upgrade voting. If they match, miner is registered as voter in favor of upgrade. In any other cases everything will work as usual.

Below you can find quickly scribbled code, that shows how it could be implemented using last byte of a memo. It is just a voting process and it doesn't have any threshold that needs to be reached in order to start freeze/warmup process but that is a minor detail. 

```lisp
(define-data-var upgrade-voting-active bool false)
(define-data-var upgrade-byte (buff 1) 0x00)
(define-data-var upgrade-nonce uint u0)
(define-map upgrade-voters
    principal  ;; who
    uint       ;; nonce
)

(define-public (activate-upgrade-voting (upgrade-byte (buff 1)))
    (begin
        ;; only contract deployer can start upgrade voting process
        (asserts! (is-eq tx-sender CONTRACT-OWNER) (err ERR-UNAUTHORIZED))
        (asserts! (not (var-get upgrade-voting-active)) (err ERR-UPGRADE-VOTING-IN-PROGRESS))
        (var-set upgrade-voting-active true)
        (ok (var-set upgrade-byte upgrade-byte))
    )
)

(define-private (process-memo (memo (buff 34)))
    (begin
        (if (var-get upgrade-voting-active)
            (if (is-eq (var-get upgrade-byte) (element-at memo (- (len memo) 1)))
                (match (map-get? upgrade-voters tx-sender)
                    voter true
                    (begin
                        (map-set upgrade-voters tx-sender (+ (var-get upgrade-nonce) u1))
                        (var-set upgrade-nonce (+ (var-get upgrade-nonce) u1))
                    )
                )
                true
            )    
            false
        )
        (print memo)
    )
)

(define-public (mine-tokens (amount-ustx uint) (memo (optional (buff 34))))
    (begin
        (if (is-some memo)
            (process-memo memo)
            none
        )
        (mine-tokens-at-block block-height (get-or-create-miner-id tx-sender) amount-ustx)
    )
)

```

**Comment by LNow** [comment]
@whoabuddy suggested here (https://github.com/citycoins/citycoin/issues/83#issuecomment-864280600) a very nice workflow how we could tackle setting up and activating trusted minting contract.

To recap his proposal:

> 1. A new contract version is created and deployed
> 2. The city wallet adds the new contract to the approved contract list in the minting contract, but as inactive status
> 3. The miners vote for the upgrade using the memo to activate the new contract
> 4. Once threshold is reached:
    * new contract enters active status in minting contract and is available
    * sunset/prepare phases begin between v1 to v2 as part of upgrade process

If we consider it as something we would like to implement it will become clear that we need a code that will handle:
 * registering minting contract candidate
 * counting votes
 * all below at once: 
     * converting minting contract candidate into actual minting contract
     * triggering sunset/freeze procedure in old contract
     * triggering sunrise/warmup procedure in new contract

And preferably be located in separate contract, as we don't want to bloat CityCoin contract more than we have to.

Because I couldn't think of a better name, I called this code **Core**.

**Core** is a separate contracts. It will allow City to register/add new version of main CityCoin contract as candidate for new trusted minting contract. Once that'll happen voting on this candidate will be open.

Miners who send some message in memo will be considered as potential voters and if selected part of their message will match a value set for active voting (see  https://github.com/citycoins/citycoin/issues/84#issuecomment-863205610), it will be counted as "yes" vote.

When we reach "consensus threshold", it will:
 * add new CityCoin contract to Token contract as trusted caller
 * trigger sunset procedure in old CityCoin contract
 * trigger sunrise procedure in new CityCoin contract
 * close voting

Bellow you can find simplified sequence diagram that presents registering new CityCoin contract in Core by City, and last Miner vote that surpass "consensus threshold" and as a results kick-start CityCoin contract upgrade. 

![citycoin-voting](https://user-images.githubusercontent.com/829210/122978728-20dfea80-d397-11eb-9a8b-5aa934739ab2.png)
<details>
  <summary>Code behind diagram</summary>

```
title DAO Voting using memo

participant Token
Actor City
participant Core
participant CityCoin

note right of City
Once CityCoin V2 is deployed
City can register it in Core.
Registering contract enables voting.
end note
City->+Core:register(CityCoin V2)
Core-->-City: ok

Actor Miner
Miner->+CityCoin: mine-tokens(100ustx, "vote yes on V2")
CityCoin->+Core: add-vote(miner, V2)

note left of Core
Core will add trusted caller
to Token only when threshold
is reached.
end note
Core->+Token: add-trusted-caller(CityCoin V2)
Token-->-Core: ok

note right of Core
sunset should be set 
N blocks in the future
end note
Core->+CityCoin: sunset()
CityCoin-->-Core: ok

note right of Core
sunrise should be set 
N blocks in the future
end note
Core->+CityCoin V2: sunrise()
CityCoin V2-->-Core: ok

Core-->-CityCoin: ok
CityCoin->-Miner: ok
``` 
</details>

If we would pick something like described above, then whole logic responsible for storing/changing City wallet could be moved into **Core**.

**Comment by whoabuddy** [comment]
I think core makes sense, and it makes me wonder if we should prefix all the naming with the citycoin name in anticipation of future city contracts, e.g.

- citycoin-token
- citycoin-core
- citycoin-v1/v2/v3
- miamicoin-token
- miamicoin-core
- miamicoin-v1/v/2/v3
- and so on...

Happy to file a separate issue for that if others agree.

---

Adding on to the voting logic in the above comment, I see a few different approaches we can take here and want to highlight/document some of the recent Discord conversations below. @LNow raised some great questions and I've provided my thoughts below each, anyone is  welcome to comment as well.

> 1) How do we want to calculate and/or set number of votes candidate must get to be activated?

I see two approaches here, we can look at a miner's version bit as a **vote** from their address, or we can look at it as an ongoing **signal** across multiple blocks. The latter may be the better approach, but will provide some diagrams in a separate comment further explaining this thought.

> 2) Do we want to restrict the time when miners can vote for a candidate (ie. 1000 blocks from the time candidate was added)?

I think this makes sense, while in theory a new version would be voted in right away, especially for a bug fix over a feature add, having an option to expire after N blocks sounds reasonable. I'd think this should be at least 1-3 months, though, to give enough time to prepare UI and raise awareness. *(thinking we should have a notify mailing list / Discord channel for updates similar to what Stacks is doing for announcements)*

> 3) Do we want to set any delay between accepting candidate and activating it?

Yes, similar to how registering in the contract starts a delay I think this makes sense as well. This could be built into the sunrise/sunset functionality, and in the case of a critical bug fix, the city wallet should be able to accelerate it if needed. Maybe 1 week by default to allow for UI upgrades and testing with the new contract (in case consensus is reached right away), and a vote from the city wallet accelerates it to activate immediately. This could be a breaking change to the UI, however that may be worth the temporary down time in the case of a critical bug.

> 4) Should the activation of an accepted candidate happen automatically or do we want City to trigger the activation or maybe we want to activate it automatically with some delay and give City the power to postpone it or speed it up?

I vote automatically, at this point a new contract was developed, the city added it to the list, miners have signaled they want to use it - at which point it goes into effect following the thoughts I expressed in 3.

> I was thinking about it for quite a bit of time, and although X % of all miners seems to be the most reasonable way for setting activation threshold, it might be difficult to reach if we'll have large number of miners right after v1 deployment, and it will drop dramatically after.
> Maybe miners should not vote when they call mine-tokens, but when they call claim-mining-reward? And instead of counting number of miners, we should count number of claimed blocks in which the miners signaled their acceptance for upgrade?
> Very similar approach is used in bitcoin. https://en.bitcoin.it/wiki/BIP_0009

This could be a viable alternate approach, but I'm still leaning toward the act of mining being the way miners signal something, rather than the act of claiming rewards. My concern is how miners can go back and claim at any time, and how to calculate a valid vote across the varied block heights they may be claiming for. Does this mean that if I don't claim for 10 block rewards, then claim those later during an upgrade process, I could tip the scale?

Not shooting it down - I just think it's something we need to explore more in-depth.

**Comment by whoabuddy** [comment]
All diagrams below are [documented in Miro](https://miro.com/app/board/o9J_lGTdbbg=/?moveToWidget=3074457360731666424&cot=14) with screenshots provided for simplicity.

If we consider the same states as BIP-0009 related to our described process:

![Screenshot from 2021-06-28 08-20-39](https://user-images.githubusercontent.com/9038904/123662294-2c457280-d7ea-11eb-8ba6-c02e85ae6b5f.png)

And if we consider the basic upgrade process we a proposing:

![Screenshot from 2021-06-28 08-20-58](https://user-images.githubusercontent.com/9038904/123662335-37989e00-d7ea-11eb-854a-6e2ad7af3a29.png)

I see two pathways we can take in using `mine-tokens` to evaluate miners votes.

## Option A - miner voting

In this scenario, a map of voters is created storing the address that voted while the upgrade process is occurring. The threshold is a set to a number of votes, and each address can submit only once during the voting period.

![Screenshot from 2021-06-28 08-21-40](https://user-images.githubusercontent.com/9038904/123662488-61ea5b80-d7ea-11eb-857f-095dcd55f4f2.png)

## Option B - miner signaling

In this scenario, miners continuously vote by signaling with each transaction, and the results within N blocks are reviewed as part of the upgrade process. If N blocks reach consensus, then the upgrade is triggered.

![Screenshot from 2021-06-28 08-22-16](https://user-images.githubusercontent.com/9038904/123662521-67e03c80-d7ea-11eb-9118-ea68170efac4.png)

---

I'm leaning toward Option 2 as we can set a threshold (% TBD) and miners have to continuously broadcast versus make a single vote. Then, if we are in the "Started" phase (defined by a `bool`), the contract can check back N blocks for consensus.

This makes sure that active miners are registering votes versus just one address submitting a mining TX once - but curious what others think and what the possible downsides are to this approach.

**Comment by whoabuddy** [comment]
Here is a (very rough) UI mock-up of how to make it easier for miners to submit a TX, based on if the contract is in the "Started" phase described above. Checking the box would insert the correct memo and disable the memo field for editing. It could also control if mine for 30 blocks is available, although I think specifying the upgrade for 1 or 30 blocks should be a valid action.

![Screenshot from 2021-06-28 08-53-33](https://user-images.githubusercontent.com/9038904/123666993-8f390880-d7ee-11eb-877c-8344a5714173.png)

**Comment by jcnelson** [comment]
I'd go with Option B.  Miner signaling is something that I think people are more familiar with, and it's more inclusive -- by requiring that miners continuously signal over a fixed window of time in order for an upgrade to happen, it gives the greater community a chance to participate.

I like the idea of having a "core" contract that acts as a registry for all versions of a particular city's CityCoins contract.  In fact, you could probably have a single "core" contract that manages all versions for all CityCoins contracts for all cities if you wanted.  So, if you wanted to create AustinCoin or the like, you'd deploy the AustinCoin contract, and use the "core" contract to vote it into the CityCoins wallet.  Is that what you were going for?

**Comment by whoabuddy** [comment]
> I like the idea of having a "core" contract that acts as a registry for all versions of a particular city's CityCoins contract. In fact, you could probably have a single "core" contract that manages all versions for all CityCoins contracts for all cities if you wanted. So, if you wanted to create AustinCoin or the like, you'd deploy the AustinCoin contract, and use the "core" contract to vote it into the CityCoins wallet. Is that what you were going for?

Not quite, I was imagining one set of contracts per city, with the operations divided between core, token, and "engine" which is the piece that can be upgraded. I like the idea of each city being its own container, in which they have their designated wallet, and can individually upgrade their contracts.

I could see having an overall registry contract or something similar to track the different city implementations, 
 city wallets, etc, especially for the UI to be more responsive or to provide additional stats, but did not want to get too far out of scope for what we are doing in the short term in trying to get one city up and running :slightly_smiling_face:

**Comment by whoabuddy** [comment]
> One thing I would change here is that instead of checking whether part of memo field corresponds to our voting magic bits, I think we should add optional version/voting field to mine-tokens function. It will simplify whole logic as instead of slicing memo, to find out if it contains vote or not, we'll just verify if version field has been filled by miner and if it has expected value.

I wish we had those native slice functions coming in 2.1... but I agree the simplest path is probably the best at this point, especially around anything that would increase execution costs. We should develop a basic standard for the versions and their lifecycles, and that's something we can turn around pretty easily for the docs once established.

We should provide a getter function for if an upgrade is active or not, with some basic details like `get-pox-lite-info` returns, so that the UI can accurately notify, represent, and submit upgrade TXs. 

> We need to think about fixed arbitrary voting period length (ie. 2016 blocks) or set is as default option, but allow City to also pick custom length (no shorter than X and no longer than Y).

There are two timelines I'm thinking of in this regard:

- **voting period:** this can remain a default value and I'm not sure it needs to be modified. I think it makes sense to synchronize with our stacking intervals of 2,100 blocks, unless 2 weeks doesn't feel like long enough for a vote to take effect. Given miners are mining either every 10 minutes or 4 hours (if continuous), it feels like there are plenty of opportunities in those two weeks to tally the votes.
- **activation period:** once consensus is reached above, how fast does the new contract go into effect, and what does the sunrise/sunset process look for the old and new contracts. I suspect we will just need to record block heights in each and use that for calculating old state or accepting new transactions, and I think we'll learn more through trying this approach.

For the activation period, I think this is where it's crucial that the city can reduce (or near eliminate) the sunrise/sunset timing if necessary. If it's a feature add, a nice-to-have, or just a general upgrade - we could lean on the 2,100 block period again for transition. If it's a critical bug fix, there should be a way to signify that and trigger an upgrade as fast as possible (once accepted/voted).

> The simplest votes counting solution that comes to my mind consists of two counters stored in a map (one entry in map = one voting period).

With this and thinking of the getter function described before, we should probably assign an ID to the voting period that can be used to look up the contract info, upgrade status, and other information in a separate `read-only` function. That way we are using as small of a map as possible in relation to the mining TX. 

> miners-count will increment with every successful call of mine-tokens
> votes-count will increment with every successful call of mine-tokens in which miner signaled his approval for upgrade

Threshold is a big question here, I'm wondering what the best way to determine consensus given the possible variations in the number of miners. Is there a threshold that would work for 3, 10, 30, 120? Percentage seems like a viable approach but I'm curious if there is a better way to handle this, or if we do something creative like set tiers.

- <5 miners = 60% (3 of 5, works for 2 of 3)
- >5, <10 miners = 70% (7 of 10, works for 5 of 7 but not 4 of 6)
- etc. etc.

The numbers above are just made up, but I bet there's a standard out there that we can lean on.

> We can also model it as additional function, that can be called by anyone to not increase mining costs.

I'm not sure if we want to open the function up outside of mining because we want the miners who are sending mining TXs to be the ones making a vote. We could see if they're already in the `miners-count` map, but that means more lookups (albeit, separate from the mining TX).

> And with regards to having "core" contract, I'm leaning more and more towards concepts similar to eternal storage, where you have one contract that act like a storage for everything and you surround it with multiple contract that consists all business logic.
For now it is just an idea in my mind, but first let's focus on implementing at least basic upgrade logic. Then we can think if we want to evolve it further or not.

This sounds great, I agree we can do a lot by trying to put the simplest pieces in place then deciding where to go from there. The hub and spoke approach definitely makes sense, especially now as we are starting to look at a number of separate contracts.

Everything else sounds great, I think the best we can do is start by implementing the basic version, and see where that takes us!

**Comment by LNow** [comment]
> I'm not sure if we want to open the function up outside of mining because we want the miners who are sending mining TXs to be the ones making a vote. We could see if they're already in the miners-count map, but that means more lookups (albeit, separate from the mining TX).

I was thinking only about triggering upgrade (starting sunrise/sunset periods) after successful voting.
Upgrade must be triggered somehow. Once we collect all votes and voting period ends it doesn't matter how or who will call a function that will verify if miners accepted upgrade or not. Miners made their decision and we can't change that. Someone has to do it, therefore I think it should be a public function callable by anyone. Miners, stackers, City, developers - anyone.
Why not add it to mining or claiming mining reward? Because it will increase complexity and execution costs.
Why not add it to stacking or claiming stacking reward? Because it will increase complexity and execution costs.

**Comment by whoabuddy** [comment]
Yes I misunderstood, I thought you were saying the public function was for the act of voting, not for processing the result / triggering the upgrade. I agree this could be one called by anyone.

**Comment by 314159265359879** [comment]
On the Dashboard could we also have a bar that shows how much (absolute and fraction?) of the citycoins are stacked? 
![image](https://user-images.githubusercontent.com/33360391/124598913-6dcac300-de65-11eb-8ed8-d24b8386b9ee.png)

**Comment by whoabuddy** [comment]
Yes we're working on some UI improvements in the `citycoins-ui` repo, see https://github.com/citycoins/citycoin-ui/issues/27 for displaying locked CC on the profile and added it to https://github.com/citycoins/citycoin-ui/issues/37

**Comment by SyAsteria** [comment]
closing #7 as it's a dup of this

**Comment by whoabuddy** [comment]
I've been thinking a lot about this and have an interesting approach I'd like input from others on.

A lot of sites that talk about contract upgrades focus on separating the storage from the logic, but in our case, the original `citycoin.clar` contract (renamed `citycoin-logic-v1`) contained the information on who mined, who stacked, and what they can claim as a result.

Rather than moving all that storage into the core contract, I think we can let each contract be its own database, with functions that are only called from the core contract rather than completely public. By restricting the caller to being the `citycoin-core` contract, we can use the core as an extra layer to ensure the correct logic is enacted such that:

- `citycoin-core` would handle the miner registration and activation of mining, since this only happens once per CityCoin
- `citycoin-core` would have functions for mining, stacking, claiming mining rewards, and claiming stacking rewards that call the currently active `citycoin-logic` contract
  - `citycoin-logic` contract would adhere to a trait that specifies it must support these 4 functions
  - `citycoin-logic` contract would only allow these public functions to be called by `citycoin-core`
- `citycoin-core` would handle the voting and upgrade process
  - `citycoin-logic` contract would adhere to a trait that specifies startup, shutdown, and contract info
    - startup would represent when mining/stacking are available in the contract
    - shutdown would represent when mining/stacking are disabled, but mining/stacking claims would still be active
    - get-contract-info would show the version and state of the contract
    - this trait could be used for the initial startup of V1 after registration
  - `citycoin-core` would track the active contract and route calls to the correct version of `citycoin-logic`
  - `citycoin-core` would ensure that only the city wallet address can submit an upgrade, and miners will still have to vote for it to become active

This also allows us to move some functions specifically to `citycoin-core`, including:

- updating the city wallet address (only available to the city wallet address as a sender)
- updating the `citycoin-token` URI
- updating the `trusted-callers` for the citycoin-token contract, ensuring each logic contract is the only one that can mint tokens via `claim-mining-reward`

The upgrade process would have to take into consideration:

- users mining over 30 blocks, if those 30 blocks extend past the shutdown block height
- users in a current or future reward cycles and unlocking of tokens
- users calling to claim mining or stacking rewards would be evaluated against the block height specified, and routed to the correct contract to claim rewards

The advantage here is any new `citycoin-logic` can support a wide array of changes to the mining and stacking functions as long as they adhere to the defined `citycoin-logic` trait, however it still requires:

- someone to develop and deploy the contract
- the city wallet to submit the contract for an upgrade
- miners to vote the contract into activation

Those extra layers should provide the needed security to prevent a malicious contract from being deployed, while still allowing for a decent amount of flexibility in how current and future contracts will operate.

A quick mock-up of the idea here as well:

![Screenshot from 2021-07-13 16-24-45](https://user-images.githubusercontent.com/9038904/125537680-72af4cfb-c56f-4f1b-90fa-d9a2f90a0241.png)

There are still some open questions around execution costs, but I'm anticipating it should be close to the current costs because there is minimal logic laid over the top of calling a function, e.g.

- `citycoin-core: mine-tokens`: check map for currently active contract
- `citycoin-core: mine-tokens`: call `mine-tokens` function in currently active contract
- `citycoin-logic: mine-tokens`: continue normal execution of mining function

or another example:

- `citycoin-core: claim-mining-reward`: evaluate submitted block height against active contract start height
  - if: `block height > start height` then call `claim-mining-reward` in active contract
  - else: find `block height > start height` from past contracts in map until match
  - then call appropriate `citycoin-logic` contract to process `claim-mining-reward`
